streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1455,6 +1455,35 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
         }
     }
 
+  /**
+   * Zips two streams together with a specified function.
+   * Use the latest element of another stream if no element is available from another stream.
+   */
+  final def zipWithLatest[R1 <: R, E1 >: E, B, C](that: ZStream[R1, E1, B])(
+    f: (A, B) => C
+  ): ZStream[R1, E1, C] = {
+    var latestLeft  = null.asInstanceOf[A]
```

Comment:
Outside of the runtime of ZIO, we try not to use bare `var`s, `null`s or casts unless absolutely necessary. Let's try and rewrite the combinator without these.

Reply:
`null` can be avoided with `Some`. I still need locally mutable variables. It seems update a `Ref` of ZIO requires extra overhead as it uses synchronization primitive. This is the reason I avoided `ref`.